src/charm.py

```diff
@@ -152,19 +143,17 @@ def _on_start(self, _: typing.Any) -> None:
         """Handle the start event."""
         self._reconcile()
 
+    @validate_config_and_integration(defer=False)
     def _on_get_certificate_action(self, event: ActionEvent) -> None:
         """Triggered when users run the `get-certificate` Juju action.
 
         Args:
             event: Juju event
         """
         hostname = event.params["hostname"]
-        tls_certificates_relation = self._tls.get_tls_relation()
-        if not tls_certificates_relation:
-            event.fail("Certificates relation not created.")
-            return
+        tls_information = TLSInformation.from_charm(self)
 
-        tls_rel_data = tls_certificates_relation.data[self.app]
+        tls_rel_data = tls_information.tls_requirer_integration.data[self.app]
```

Comment:
Just checking in: could this raise a `KeyError` if the self.app does not exist here?

Reply:
What I've done previously for integration data was to model them and initialize the DTO like so:
https://github.com/canonical/github-runner-operator/blob/main/src/charm_state.py#L771-L793
(would this be helpful here?)

Reply:
Thanks, this is a valid point, I mentioned it in the discussion started by Greg above, so I'm marking this as resolved as well.